src/heatmap.ts

```diff
@@ -0,0 +1,761 @@
+import { GradShader, ColorShader, ImageShader } from "./shaders";
+import {
+  BackgroundImageConfig,
+  GradientElement,
+  HearmapExData,
+  HeatmapConfig,
+  MappedGradient,
+  Point,
+  ShaderProgram,
+  Translate,
+} from "./types";
+import {
+  createImageShader,
+  createGradiantShader,
+  createColorShader,
+} from "./utils/shaderUtils";
+import {
+  isNullUndefined,
+  isNotNumber,
+  isSortedAscending,
+  getPixelRatio,
+} from "./utils/utils";
+
+function gradientMapper(grad: GradientElement[]): MappedGradient {
+  if (!Array.isArray(grad) || grad.length < 2) {
+    throw new Error(
+      "Invalid gradient: Expected an array with at least 2 elements."
+    );
+  }
+
+  if (!isSortedAscending(grad)) {
+    throw new Error("Invalid gradient: Gradient is not sorted");
+  }
+
+  const gradLength = grad.length;
+  const values = new Float32Array(gradLength * 4);
+  const offsets = new Array(gradLength);
+
+  grad.forEach(function (d, i) {
+    const baseIndex = i * 4;
+    values[baseIndex] = d.color[0] / 255;
+    values[baseIndex + 1] = d.color[1] / 255;
+    values[baseIndex + 2] = d.color[2] / 255;
+    values[baseIndex + 3] = d.color[3] !== undefined ? d.color[3] : 1.0;
+    offsets[i] = d.offset;
+  });
+
+  return {
+    value: values,
+    length: gradLength,
+    offset: offsets,
+  };
+}
+
+function extractData(data: Point[], self: HeatmapRenderer): HearmapExData {
+  const len = data.length;
+  //   TODO ISSUE?
+  let { posVec = [], rVec = [] } = (self.hearmapExData || {}) as HearmapExData;
```

Comment:
And what should be types for posVec and rVec? Are they just normal arrays or Float32Array? The reason for confusion is that some places it is being used as arrays and some other places as Float32Array.

Reply:
they are of type Float32Array. please correct the code, make sure we use float32 at all places.